### ISSUE TITLE
checkpoint之后恢复，不知道怎么用新的参数覆盖checkpoint之前的参数

### ISSUE THREAD

**Issue by 123wula** [ultralytics/ultralytics#21499]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.
我中断训练之后无法使用新的参数进行覆盖中断之前的参数

这是中断之前的参数hyp_custom.yaml：
`
epochs: 300
patience: 100
batch: 8
imgsz: 640
save: true
save_period: -1
cache: false
device: '0'
workers: 2

optimizer: SGD

iou: 0.5

save_txt: true

lr0: 0.04
lrf: 0.01

`
这是新的参数表
`
data: data.yaml
task: detect
mode: train
model: runs\detect\train2\weights\last.pt

epochs: 300
patience: 100
batch: 16
imgsz: 640
save: true
save_period: -1
cache: false
device: '0'
workers: 2
project: null
name: train2

optimizer: SGD

cos_lr: true
resume: true
amp: true

iou: 0.5

save_txt: false

lr0: 0.0025
lrf: 0.01

`

### Question

1. 使用命令行

yolo task=detect mode=train resume model=runs/detect/train2/weights/last.pt  batch=16

2. 使用python api

`
def main():
    # 加载预训练模型
    model = YOLO('../datasets/runs/detect/train/weights/last.pt')
    model.cfg='hyp_custom.yaml'
    results = model.train()

if __name__ == '__main__':
    # Windows多进程
    mp.freeze_support()  # 添加这一行
    main()
`
或者
`
def main():
    # 加载预训练模型
    model = YOLO('../datasets/runs/detect/train/weights/last.pt')
    results = model.train(cfg='hyp_custom.yaml')

if __name__ == '__main__':
    # Windows多进程
    mp.freeze_support()  # 添加这一行
    main()
`
甚至使用cfg='../datasets/runs/detect/train/args.yaml'(这里args.yaml进行修改)
甚至修改default.yaml
都没有用，我发现只要加载了../datasets/runs/detect/train/weights/last.pt，参数就不会再更新

**Comment by UltralyticsAssistant** [comment]
👋 Hello @123wula, thank you for reaching out to Ultralytics 🚀! We appreciate the detailed description of your workflow. For new and returning users, we recommend reviewing the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as answers to many common questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate your issue more efficiently.

If this is a custom training ❓ Question, please share as much detail as possible. This includes dataset image examples, training logs, and confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

You’re invited to join the Ultralytics community! For quick chat and support, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For more in-depth discussions, try [Discourse](https://community.ultralytics.com/). Or connect with others on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This helps verify your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. These tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu, every 24 hours and on every commit.

---

This is an automated response to help streamline support. An Ultralytics engineer will review your question and assist you soon. Thank you for your patience! 🛠️

**Comment by Y-T-G** [comment]
If you use `cfg` to pass the YAML, it should work

**Comment by 123wula** [comment]
Thank you for help,actually I  want to know ,has anyone changed args when resuming ,or tips for observing args' behavior.Maybe I have to read codes..
I did as follows:
yolo task=detect mode=train resume model=runs/detect/train2/weights/last.pt batch=16 cfg=hyp_custom.yaml
or
results = model.train(cfg='hyp_custom.yaml')#in run.py
or 
model.cfg='hyp_custom.yaml'
results = model.train()
all failed!
The modified args.yaml in the same root with weights will be recoverd again
I triggered a error but found little...only 'batch' can get changed in cmd (windows11)

**Comment by glenn-jocher** [comment]
The issue you're experiencing is that when using `resume=True`, YOLO loads the complete training state from the checkpoint including the original hyperparameters, which overrides your new parameters.

Here's the solution - instead of using `resume=True`, load the checkpoint weights as a pretrained model and start a new training session:

**Command Line:**
```bash
yolo train data=data.yaml model=runs/detect/train2/weights/last.pt batch=16 lr0=0.0025 cos_lr=True
```

**Python API:**
```python
from ultralytics import YOLO

model = YOLO('runs/detect/train2/weights/last.pt')  # Load weights only
results = model.train(
    data='data.yaml',
    batch=16,
    lr0=0.0025,
    cos_lr=True,
    # Add other parameters you want to change
)
```

The key difference is removing `resume=True`. This way you load the trained weights but start a fresh training session with your new hyperparameters. The training will continue from where the weights left off but with your new configuration applied.

**Comment by 123wula** [comment]
@glenn-jocher Thank,I have tried!
Yes, all things work well from scratch, but I wonder whether my original setting of the SGD optimizer would be affected. What about the learning rate scheduler?

**Comment by glenn-jocher** [comment]
When you load weights without `resume=True`, the optimizer state (including SGD momentum buffers) and learning rate scheduler state are reset and start fresh with your new configuration. This is actually the correct behavior when you want to change hyperparameters.

The model weights are preserved, but the optimizer will initialize with your new `lr0=0.0025` and `cos_lr=True` settings. The SGD optimizer will start building new momentum buffers from scratch, and the cosine learning rate scheduler will begin its schedule from epoch 0 with your new learning rate.

This is typically what you want when changing training parameters - you keep the learned features in the model weights but allow the optimizer to adapt to the new hyperparameters you've specified.

**Comment by 123wula** [comment]
Thanks so much for your detailed response—it’s incredibly helpful!

**Comment by glenn-jocher** [comment]
You're welcome! Glad I could help clarify the difference between resuming training state versus loading weights for a fresh training session. Feel free to reach out if you have any other questions!
